src/main/java/org/betonquest/betonquest/objectives/BlockObjective.java

```diff
@@ -28,42 +30,64 @@ public class BlockObjective extends CountingObjective implements Listener {
         super(instruction);
         selector = instruction.getBlockSelector();
         exactMatch = instruction.hasArgument("exactMatch");
         targetAmount = instruction.getVarNum();
         noSafety = instruction.hasArgument("noSafety");
+        final String stringlocation = instruction.getOptional("loc", "false");
+        location = "false".equalsIgnoreCase(stringlocation) ? null : new CompoundLocation(instruction.getPackage(), stringlocation);
+        hasLocation = !"false".equalsIgnoreCase(stringlocation);
+        ignorecancel = instruction.hasArgument("ignorecancel");
+
     }
 
     @Override
     public String getDefaultDataInstruction(final Profile profile) {
         return String.valueOf(targetAmount.getInt(profile));
     }
 
-    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
-    public void onBlockPlace(final BlockPlaceEvent event) {
+    @EventHandler(priority = EventPriority.HIGHEST)
+    public void onBlockPlace(final BlockPlaceEvent event) throws QuestRuntimeException {
         final OnlineProfile onlineProfile = PlayerConverter.getID(event.getPlayer());
         if (containsPlayer(onlineProfile) && selector.match(event.getBlock(), exactMatch) && checkConditions(onlineProfile)) {
+            if (hasLocation && !event.getBlock().getLocation().equals(location.getLocation(onlineProfile))) {
+                return;
+            }
             if (getCountingData(onlineProfile).getDirectionFactor() < 0 && noSafety) {
                 return;
             }
-            handleDataChange(onlineProfile, getCountingData(onlineProfile).add());
+            handleDataChange(onlineProfile, getCountingData(onlineProfile).add(), event.isCancelled());
         }
+
     }
 
-    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
-    public void onBlockBreak(final BlockBreakEvent event) {
+    @EventHandler(priority = EventPriority.HIGHEST)
+    public void onBlockBreak(final BlockBreakEvent event) throws QuestRuntimeException {
         final OnlineProfile onlineProfile = PlayerConverter.getID(event.getPlayer());
         if (containsPlayer(onlineProfile) && selector.match(event.getBlock(), exactMatch) && checkConditions(onlineProfile)) {
+            if (hasLocation && !event.getBlock().getLocation().equals(location.getLocation(onlineProfile))) {
+                return;
+            }
             if (getCountingData(onlineProfile).getDirectionFactor() > 0 && noSafety) {
                 return;
             }
-            handleDataChange(onlineProfile, getCountingData(onlineProfile).subtract());
+            handleDataChange(onlineProfile, getCountingData(onlineProfile).subtract(), event.isCancelled());
+
         }
     }
 
-    private void handleDataChange(final OnlineProfile onlineProfile, final CountingData data) {
+    private void handleDataChange(final OnlineProfile onlineProfile, final CountingData data, final boolean cancel) {
+        if (cancel && !ignorecancel) {
```

Comment:
In my opinion this check should be first because it is the most simple one.

Reply:
I originally had that condition inside the Events itself.
When i had that, maven verify threw PMD violations for too high complexity of the code. (Idek why, it's a single if-clause. No matter where i put it in that method, it gave me the violation)

Reply:
Other ways to reduce complexity are for example to move combined conditions out into their own methods (if it actually makes sense):
```java
// before
if (a.equals(b) && (b.val == f.getV() || b.val * 2 < other.getMax())) {
  // do stuff
  }
  // after
  if (isRelevantForObjective(b)) {
    // do stuff
  }
```

But please try not to change the code to please the PMD rules, but instead improve the code for humans and PMD will usually like it. Especially don't take the worse option because it is easier to "fix PMD". It is meant as a tool to warn if some piece of code should be improved, not to punish. If worst comes to worst we can always decide to suppress PMD warnings (but that should not be necessary here).

Reply:
I did make it now visually pleasing and PMD conform.